### PR TITLE
feat: enable customized group inverse

### PR DESCRIPTION
**Pull request by chokobole**
# Description

Previously, `Negative()` and `Inverse()` depends on `NegInPlace()` and `InverseInPlace()` internally. But there's a case that those functions require the return type to be different. In this case, the derived class needs to define its own `Negative()` and `Inverse()`.

### CODE REVIEW

**Comment by TomTaehoonKim** [comment]
There is a typo in the first commit message! Fix assignemnt -> assignment please!

File: tachyon/math/base/groups.h [hunk 1]
```diff
@@ -48,7 +48,7 @@ class MultiplicativeGroup : public MultiplicativeSemigroup<G> {
   template <
       typename G2,
       std::enable_if_t<internal::SupportsDivInPlace<G, G2>::value>* = nullptr>
-  constexpr auto& operator/=(const G2& other) {
```
Comment (TomTaehoonKim) [inline tachyon/math/base/groups.h:51]: What other types could come if we haven't forced to return `G&`?

File: tachyon/math/base/groups.h [hunk 2]
```diff
@@ -96,23 +97,23 @@ class AdditiveGroup : public AdditiveSemigroup<G> {
...
-  constexpr G& operator-=(const G2& other) {
-    G* g = static_cast<G*>(this);
-    return g->SubInPlace(other);
-  }
-
-  // Subtraction in place: a += (-b)
-  template <
-      typename G2,
-      std::enable_if_t<!internal::SupportsSubInPlace<G, G2>::value &&
+      std::enable_if_t<internal::SupportsSubInPlace<G, G2>::value ||
```
Comment (TomTaehoonKim) [inline tachyon/math/base/groups.h:111]: Can you fix the commit message? I think you have united `operator-=()` not `operator*=()`.

File: tachyon/math/base/groups_unittest.cc [hunk 3]
```diff
@@ -57,6 +57,39 @@ TEST(GroupsTest, DivOverMul) {
...
+    explicit IntInverse(int denominator) : denominator_(denominator) {}
+    IntInverse(const IntInverse& other) : denominator_(other.denominator_) {}
+
+    bool operator==(const IntInverse& other) const {
+      return denominator_ == other.denominator_;
+    }
+
+   private:
+    int denominator_;
+  };
```
Comment (TomTaehoonKim) [inline tachyon/math/base/groups_unittest.cc:73]:
```suggestion
  class IntInverse {
   public:
    IntInverse() = default;
    explicit IntInverse(int denominator) : denominator_(denominator) {}
    IntInverse(const IntInverse& other) : denominator_(other.denominator_) {}

    bool operator==(const IntInverse& other) const {
      return denominator_ == other.denominator_;
    }

   private:
    int denominator_ = 0;
  };
```
How about like this as we used to do?

File: tachyon/math/base/groups_unittest.cc [hunk 4]
```diff
@@ -57,6 +57,39 @@ TEST(GroupsTest, DivOverMul) {
...
+    explicit Int(int value) : value_(value) {}
+    Int(const Int& other) : value_(other.value_) {}
+
+    IntInverse Inverse() const { return IntInverse(value_); }
+
+    bool operator==(const Int& other) const { return value_ == other.value_; }
+
+   private:
+    int value_;
+  };
```
Comment (TomTaehoonKim) [inline tachyon/math/base/groups_unittest.cc:87]:
```suggestion
  class Int : public MultiplicativeGroup<Int> {
   public:
    Int() = default;
    explicit Int(int value) : value_(value) {}
    Int(const Int& other) : value_(other.value_) {}

    IntInverse Inverse() const { return IntInverse(value_); }

    bool operator==(const Int& other) const { return value_ == other.value_; }

   private:
    int value_ = 0;
  };
```

File: tachyon/math/base/groups_unittest.cc [hunk 5]
```diff
@@ -109,4 +142,37 @@ TEST(GroupsTest, SubOverAdd) {
...
+    explicit Int(int value) : value_(value) {}
+    Int(const Int& other) : value_(other.value_) {}
+
+    IntNegative Negative() const { return IntNegative(-value_); }
+
+    bool operator==(const Int& other) const { return value_ == other.value_; }
+
+   private:
+    int value_;
+  };
```
Comment (TomTaehoonKim) [inline tachyon/math/base/groups_unittest.cc:172]:
```suggestion
  class Int : public AdditiveGroup<Int> {
   public:
    Int() = default;
    explicit Int(int value) : value_(value) {}
    Int(const Int& other) : value_(other.value_) {}

    IntNegative Negative() const { return IntNegative(-value_); }

    bool operator==(const Int& other) const { return value_ == other.value_; }

   private:
    int value_ = 0;
  };
```

File: tachyon/math/base/groups.h (hunk 1 above)
Comment (chokobole) [reply tachyon/math/base/groups.h:51]: Only `G&` can come. 

```c++
T a;
a += b; // a doesn't change its type!
```

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**